### ISSUE TITLE
Package Dependency Resolution

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1618]
**Reported by choeger on 4 Dec 2014 16:22 UTC**
# Package Depencency Resolution

This is a followup to #573 and focusses on the specification of a correct algorithm for the resolution of dependencies depending on a `uses()` annotation with multiple versions.

## Preface

The proposal for a more modular package management allowed for a set of compatible versions in a `uses()` annotation. 
Discussion of this feature revealed a need to define the process of resolving the dependencies of a given class given such an annotation. This ticket discusses such an algorithm under the following assumptions:

1. for all available packages, the name, version and uses annotation are available (e.g. by means of a pre-created index inside a package manager)

### Requirements

2. The goal is to select a subset of all available packages in a way that solves all dependencies transitively. I.e. if one package A is selected and declares usage of another package B then a version of B is also selected that has been declared compatible by A. 
3. packages are to be selected _exclusively_, i.e. if package foo exists in versions 1.0.0 and 2.0.0 either one of them may be selected but not both.  

## Input

The input is defined as a set of usage-clauses:

```
usage ::= name version USES { name {version} }
```

i.e. a usage clause is defined by the name of the annotated package and a (possible empty) list of dependencies which are given by a name and a set of compatible versions. The definition of name and version does not matter for the algorithm and are left out.

## SAT Problem

A (boolean) satisfiability problem or short SAT is defined by a set of boolean variables x_1 .. x_n and a boolean term:

```
t ::= x    //variable 
    | t * t  //conjunction ('and')
    | t + t  //disjunction ('or')
    | t -> t //implication 
    | t X t  //exclusive or
```

Solving a SAT problem `t` means to find a mapping of x_i to `true | false` s.t. `t` evaluates to `true` using boolean logic.

SAT is known to be NP-complete in general, but high-quality solvers exist that make it a practical tool. 

# General Concept (Executive Overview)

By translation of the dependency graph into a SAT problem, it is possible to generate an exact and correct solution. The following basic blocks are to be used:

when a package A in version V requires a package B in version W we can express this by the implication:

```
A_V_selected -> B_W_selected
```

i.e. when A.V shall be selected then this is only possible when B.W is also selected.

Multiple versions Wi can be expressed by OR-ing them:

```
A_V_selected -> B_W_1_selected + B_W_2_selected + B_W_3_selected
```

Multiple requirements can be composed using conjunction. Especially the goal of the process (i.e. "package A.V shall be selected") can be added:

```
A_V_selected *
(A_V_selected -> B_W_1_selected + B_W_2_selected + B_W_3_selected)
```

### Formula Generation
Given a uses clause 

```
<annotated> <version> USES { <dependency_i> { <version_i_j> } }
```

* If <annotated>_<version> has not been visited, then:

* We mark the <annotated>_<version> as visited.

* When `annotated` shall be selected, we add the formula with:

```
<annotated>_<version>
```

In any case, we create the following boolean formula:

```
<annotated>_<version> -> (<dependency_1>_<version_1_1> X ... X <dependency_1>_<version_1_n>)
                       * (<dependency_2>_<version_2_1> X ... X <dependency_2>_<version_2_m>)
                       * ...
                       * (<dependency_k>_<version_k_1> X ... X <dependency_k>_<version_k_q>)
```

* For each `<dependency_i>_<version_i_j>`, we proceed recursively on the corresponding uses clause.

### Dependency Resolution

 * Apply formula generation to the use-clauses of all packages that shall be selected (user input) 
 * join the generated formulas with conjunction (they are 'ANDed')
 * find a mapping of the visited <name>_<version> pairs to boolean variables that solves the generated formula
 * Each <name>_<version> that is mapped to `true` is selected 

## Discussion

This algorithm fulfills the requirements by construction of the formula (Proof sketch: if <name> <version> is selected it corresponding boolean variable is mapped to true, hence the right-hand side of the formula must also be true. This right-hand side consists of the dependencies).

The usage of `X` (exclusive OR) for multiple versions is both necessary and cumbersome. It is necessary to express that we do not want to install multiple versions of a package (otherwise the trivial solution would be to install _all_ packages). On the other hand, it complicates the SAT problem considerably: A lot of easy-to-use SAT solvers expect CNF (conjunctive normal form). It is possible to compile each formula into CNF, but the the size of a CNF formula is exponential in the number of XOR-ed elements inside a XOR-clause. Hence a package which is compatible to 10 or more versions would yield a CNF of 2^9 clauses.

The solution is to either use a SAT solver that allows for XOR inside formulas or to limit the amount of compatible versions a library might (or should) declare.

## Example

The following example was discussed in the Design Group:

```mo
package A ... version="1.0.0" ... uses(B(version={"1.9.0", "2.0.0"})) uses(C(version={"3.0.0"}))  ... end A;
package C ... version="3.0.0" ... uses(B(version="1.9.0")) ... end C;
```

In that case when we want to select "A_1.0.0", we expect the packages "C_3.0.0" and "B_1.9.0" to be also selected.

This can be implemented using the [cryptosat] tool, which allows for xor in CNF formulae:

```
p cnf 7 7
c Modelica package dependency example, formulas in CNF + XOR extension
c variable names:
c a__1_0_0 = 1
c a__1_0_0__sat__b = 2
c a__1_0_0__sat__c = 3
c
c b__1_9_0 = 4
c b__2_0_0 = 5
c
c c__3_0_0 = 6
c c__3_0_0__sat__b = 7
c
c "goal: install a"
1 0
c a requires both b and c
-1 2 0
-1 3 0
c "a requires b 1.9.0 or 2.0.0"
4 5 -2 0
c
6 -3 0
-6 7 0
-7 4 0
x4 5 0
```

The solution process yields:

```
s SATISFIABLE
v 1 2 3 4 -5 6 7 0
```

Meaning that variable 5 is mapped to false, i.e B_2.0.0 is **not** selected, as expected.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1618

**Comment by modelica-trac-importer** [comment]
**Comment by atroy on 4 Dec 2014 18:16 UTC**
Nice work! However, I think you should also clarify what happens if the dependency graph contains cycles. Either you check if the dependency graph is acyclic before you start the formula generation or you should explicitly cover this case.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 4 Dec 2014 19:09 UTC**
Replying to [comment:1 atroy]:
> Nice work! However, I think you should also clarify what happens if the dependency graph contains cycles. Either you check if the dependency graph is acyclic before you start the formula generation or you should explicitly cover this case.

The dependency graph will indeed be cyclic in the general case. That is why the formula generation marks visited package versions and skips them. Thus, the generation always traverses a spanning tree of the dependency graph.

**Comment by modelica-trac-importer** [comment]
**Comment by atroy on 4 Dec 2014 19:26 UTC**
Yes, but in that case I'd prefer some diagnostics telling me that the library dependencies are "fishy", instead of silently proceeding to the SAT solver. Though you could consider this as a tool issue.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 5 Dec 2014 16:03 UTC**
Now if only the rest of the spec were so unambiguously specified. :-)

Are we explicitly disallowing this?:

```mo
package A ... version="1.0.0" ... uses(B(version={"1.9.0", "2.0.0"})) uses(C(version={"3.0.0"}))  ... end A;
package C ... version="3.0.0" ... uses(B(version="2.1.0")) ... end C;
```

Notice that there is no version of package B that satisfies both A and C.

Given that we can simply consider different versions of B to be different libraries (e.g. by internally renaming them to B_1_9_0, B_2_0_0, etc.) I don't see why the case I show shouldn't be allowed.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 5 Dec 2014 16:35 UTC**
Replying to [comment:4 stefanv]:

> Given that we can simply consider different versions of B to be different libraries (e.g. by internally renaming them to B_1_9_0, B_2_0_0, etc.) I don't see why the case I show shouldn't be allowed.

Yes, we forbid that (rather let the dependency resolution fail). The reason is two-fold:

 * Modelica has no means to disambiguate names at the version level. If both `C` and `A` communicate over a function `B.f` or a connector `B.c` there is no way to express that they may come from different versions of `B`, hence there is the risk of silently introducing errors.
 * The second reason is that the dependency analysis would become pointless: If we allow multiple concurrent versions of a library to be selected, we could as well select **all** libraries and trivially fulfill any dependency requirements. Of course, this would neglect the very reason to introduce packages and libraries in the first place (modularization and scalability).

**Comment by modelica-trac-importer** [comment]
**Comment by henrikt on 8 Dec 2014 06:50 UTC**
What happened to the discussion about semantic versions?  If we would go with semantic versions, shouldn't 2.1.0 be the selected version of B in this case (since it will work as a replacement for 2.0.0)?  In particular, if we are considering semantic versioning, shouldn't we get away from the idea of internally renaming libraries to un-versioned copies with obfuscated version information in the package name?

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 8 Dec 2014 09:17 UTC**
Replying to [comment:6 henrikt]:


> What happened to the discussion about semantic versions?  If we would go with semantic versions, shouldn't 2.1.0 be the selected version of B in this case (since it will work as a replacement for 2.0.0)?

That discussion has nothing to do with this ticket. The selection of 2.0.0 over 2.1.0 is driven by the explicit requirements in `package C`. The maintainer of C has not yet approved his package for version 2.1.0. Additionally, even if 2.1.0 was assumed to replace 2.0.0, it might contain bugs that the maintainer of C noticed. Hence the version might have been explicitly removed.

> In particular, if we are considering semantic versioning, shouldn't we get away from the idea of internally renaming libraries to un-versioned copies with obfuscated version information in the package name?

I do not completely understand your request. If this is about having version annotations in the names of classes, I am strictly against this: Either you end up with an open interval annotation which basically tells you nothing or with a discrete set of compatible versions and a hell of a maintenance problem.

**Comment by modelica-trac-importer** [comment]
**Comment by henrikt on 8 Dec 2014 13:21 UTC**
Replying to [comment:7 choeger]:
> Replying to [comment:6 henrikt]:
> 
> 
> > What happened to the discussion about semantic versions?  If we would go with semantic versions, shouldn't 2.1.0 be the selected version of B in this case (since it will work as a replacement for 2.0.0)?
> 
> That discussion has nothing to do with this ticket. The selection of 2.0.0 over 2.1.0 is driven by the explicit requirements in `package C`. The maintainer of C has not yet approved his package for version 2.1.0. Additionally, even if 2.1.0 was assumed to replace 2.0.0, it might contain bugs that the maintainer of C noticed. Hence the version might have been explicitly removed.

To me the two discussions seem related, for what would be the real benefit of using semantic versioning be if it was not incorporated in the `uses` annotation's `version`?  If the library uses semantic versioning, it would be an error in the library if 2.1.0 could not be used as a substitute for 2.0.0.  It would not be the responsibility of the developer of package `A` to verify this.

> > In particular, if we are considering semantic versioning, shouldn't we get away from the idea of internally renaming libraries to un-versioned copies with obfuscated version information in the package name?
> 
> I do not completely understand your request. If this is about having version annotations in the names of classes, I am strictly against this: Either you end up with an open interval annotation which basically tells you nothing or with a discrete set of compatible versions and a hell of a maintenance problem.

I just mean that if MA would start to use semantic versioning for its libraries, a library version would be more than just a string that makes a library unique among all other libraries with the same package name.  Package `C` version 2.1.0 would actually have a lot more to do with package `C` version 2.0.0 than it has to do with package `D` version 2.1.0 — this aspect is missed when renaming them _C_2_0_0_ etc.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 8 Dec 2014 15:57 UTC**
All,

  I have two comments.  First, I am currently trying to implement dependency resolution in the impact package manager.  I'm going to try an algorithm that is simpler than Christoph's to see if I can get it to work.  I'm not sure if it is equivalent to Christoph's or not.  I'm not even sure it is a valid algorithm.  But I'm going to implement it to see if our version handling can utilize something less complicated than a full blown SAT solver and still given decent performance and a deterministic answer.  I'll post my results on a branch of the impact repository when/if I get it working.

  I also wanted to comment on the relationship to semantic versioning.  Something that we did not mention during the discussion but might be worth talking about is how strongly we take the claims in semantic versioning.  What I mean by that is that the semantic version spec implies that all non-major releases should be backward compatible.  My plan for impact is to perform dependency resolution using only explicit 'uses' clauses.  But if that fails, I think it is reasonable to make a second (less strict) pass where we assume the library authors have versioned their libraries according to the semantics of semantic versioning.  In other words, to implicitly assume that only major versions are significant when considering compatibility and then see if a solution can be found using these less strict constraints.

  That is just an update from my end.  I'm hoping to have something to show for this in the next few days.

**Comment by modelica-trac-importer** [comment]
**Comment by leo&#x2e;gall on 8 Dec 2014 16:26 UTC**
Replying to [comment:9 mtiller]:
> But if that fails, I think it is reasonable to make a second (less trict) pass where we assume the library authors have versioned their libraries according to the semantics of semantic versioning.  In other words, to implicitly assume that only major versions are significant when considering compatibility and then see if a solution can be found using these less strict constraints.

Great idea! That's exactly how a user would work, if he has to (manually) search for a compatible version.

I didn't really attempt to understand Christoph's approach, but my feeling is that there are going to be too many "bad" library versions (meaning version names not following semantic versioning) that a full dependency solver is too much for a problem which needs to be sorted out semi-automatically in most cases. Nevertheless, the more tool support, the better.

**Comment by modelica-trac-importer** [comment]
**Comment by henrikt on 9 Dec 2014 15:05 UTC**
It would make sense if a package had to declare explicitly that it uses semantic versioning, like:
```mo
package C ... version="3.1.42" semanticVersion=true ... end C;
```
Then dependency resolution would only be allowed to search for exact matches unless a package says `semanticVersion=true`.  (Note that the `uses` annotation would be the same in both cases, the difference is only that a package with `semanticVersion=true` would match more `uses` annotations.)

(By the way, only comparing based on major versions is not quite the same as using semantic versioning; that version 2.1.0 can be used as a substitute for 2.0.0 doesn't make 2.0.0 a valid substitute for 2.1.0.)

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 9 Dec 2014 15:11 UTC**
Replying to [comment:11 henrikt]:
> Then dependency resolution would only be allowed to search for exact matches unless a package says `semanticVersion=true`.

Why? We have conversion scripts already today. And some of them have empty conversion from that old version, which means they are compatible.

**Comment by modelica-trac-importer** [comment]
**Comment by henrikt on 9 Dec 2014 15:22 UTC**
Replying to [comment:12 sjoelund.se]:
> Replying to [comment:11 henrikt]:
> > Then dependency resolution would only be allowed to search for exact matches unless a package says `semanticVersion=true`.
> 
> Why? We have conversion scripts already today. And some of them have empty conversion from that old version, which means they are compatible.

There will always be things that cannot be expressed with conversion annotations, but still invalidates a new version as an automatic replacement for its predecessor.  Semantic versioning is a clean way of expressing this, since the new version would get a bumped major version.

Semantic versioning of a package would also make it less cumbersome to maintain the conversion annotations, as it would be enough to have conversion annotations for converting between major versions.

**Comment by modelica-trac-importer** [comment]
**Comment by atroy on 16 Dec 2014 14:42 UTC**
I think the strength in Christoph's algorithm is that he maps the problem of dependency resolution to a well understood problem in formal logic. If there is a consistent solution for a set of library dependencies, the algorithm is guaranteed to find a solution.

The goal of a SAT solver is to find a valid solution in order to proof that a logical formula is satisfiable. In general this solution is neither minimal (more libraries might be loaded than necessary) nor unique (in different tools different libraries might be selected). 

In the Design Meeting we've discussed an example which didn't have an optimal solution w.r.t. the order imposed by semantic versions, so I think it is okay that Christoph's algorithm doesn't consider this order. In my opinion a Modelica tool should provide a way to decide which consistent set of libraries he prefers.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 16 Dec 2014 15:15 UTC**
All.  I have also implemented an algorithm to resolve dependencies.  This particular algorithm is part of the `impact` package manager.  As such, it is not part of any specific proposal to the MA.  But it is an example of how dependencies could be resolved.  I've gotten quite busy on some other projects so the dependency resolution algorithm has not been thoroughly tested yet.  So, for the time being, it exists only on the `dependencies` branch.  The algorithm is essentially a recursive search algorithm.  You can find it at:

https://github.com/xogeny/impact/blob/dependencies/deps/index.go#L165

The basic idea is that you start with a list of libraries to resolve, e.g. `A`, `B` and `C`.  First the algorithm tries various versions of `A`.  If `A` has dependencies (e.g., `X` and `Y`) those are prepended the list of libraries yet to be resolved.  So the outer most loop of the search starts with the different possible versions of `A` and continues then with `A`'s dependencies (`X` and `Y`) and then the remaining libraries (`B` and `C`).

This effectively turns into a nested set of loops over the different versions, e.g.,


```
for verA in A.versions {
  for verX in X.versions {
    for verY in Y.versions {
      for verB in B.versions {
        for verC in C.versions {
        }
      }
    }
  }
}
```


If at any level in this for loop, it is determined that the values chosen so far (e.g., `verA=2.0.1`, `verX=1.0.5`, `verY=0.7.2`) are not explicitly compatible (according to uses annotations), then we back "prune the tree" and don't consider any combinations of other libraries by performing a `continue` on the current loop.

The versions (e.g., `A.versions`) are ordered according to most recent (using semantic versioning 2.0.0).  The first time we have chosen a version for all libraries that is consistent with the all `uses` annotations, we simply return that as the solution.  In other words, we take the first solution that we find.  But we search in such a way that we give precedence to the first libraries in the list (e.g., preferring `A` over `B` and `C`) and preference over more recent versions.

Christoph has rightfully pointed out that such an algorithm is not very elegant and I completely agree with him that you can find cases (albeit contrived, in my opinion) where you need to actually search the entire space of possibilities and the first compatible solution will be the last one.  However, such problems are rare because they depend on nearly everything depending on everything else AND the libraries being ordered in most tolerant (most `uses` annotations) to least tolerant.  I just don't see this as a common pattern.  **Furthermore** we added a test case in `impact` where there were **14** libraries with dependencies to exactly force an exhaustive search of every possibility and it still resolved the dependencies in 646 **milliseconds** on my MacBook Air.  So the bottom line is that I think this algorithm is sufficient for all practical cases.

To my mind, this algorithm has the advantage that it is

    * Relatively easy to explain
    * Relatively easy to implement
    * Deterministic
    * Yields a collection of dependencies that would be easy to explain to an engineer

I do not claim it is the best performing algorithm, but it seems to be "good enough".  I further admit that there may be some theoretical weakness somewhere in here and I am very open to a more rigorous analysis.  If a better algorithm is found that is also reasonably easy to implement without having to pull in a lot of external solver libraries, I'd be open to that as well.  My goal here was to have something that I could explain without having to resort to symbolic logic. :-)

**Comment by henrikt-ma** [comment]
For anyone interested now that this has been closed, consider contributing to #1728 instead of reopening.
